### PR TITLE
[WIP] cellVoice stub

### PR DESCRIPTION
**Pull request by isJuhn**
Dead Space: Ignition [NPUB30269] waits for the `CELLVOICE_EVENT_SERVICE_ATTACHED` before booting which is why it was stuck in nothing. This PR just adds the service attached and detached events and some error checking.

Nothing -> Ingame
image

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 1]
```diff
@@ -1,27 +1,108 @@
...
+	// axed from cellAudio
+	vm::var<sys_event_queue_attribute_t> attr;
+	attr->protocol = SYS_SYNC_FIFO;
+	attr->type = SYS_PPU_QUEUE;
+	attr->name_u64 = 0;
+
+	for (u64 i = 0; i < 100; i++)
+	{
+		// Create an event queue "bruteforcing" an available key
+		const u64 key_value = 0x80004d494f323221ull + i;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: This key is unique to cellAudioCreateNotifyEventQueue (is coming from the firmware).

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 2]
```diff
@@ -1,27 +1,108 @@
...
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	// axed from cellAudio
+	vm::var<sys_event_queue_attribute_t> attr;
+	attr->protocol = SYS_SYNC_FIFO;
+	attr->type = SYS_PPU_QUEUE;
+	attr->name_u64 = 0;
+
+	for (u64 i = 0; i < 100; i++)
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: According to the firmware, unlike cellAudioCreateNotifyEventQueue, it only tries 10 times until it runs out of keys.

File: rpcs3/Emu/Cell/Modules/cellVoice.h [hunk 3]
```diff
@@ -89,3 +89,52 @@ enum
...
+	be_t<s32> sampleRate;
+};
+
+struct CellVoicePortParam
+{
+	be_t<s32> portType;
+	be_t<u16> threshold;
+	be_t<u16> bMute;
+	be_t<float> volume;
+	union {
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.h]: brace on newline.

File: rpcs3/Emu/Cell/Modules/cellVoice.h [hunk 4]
```diff
@@ -89,3 +89,52 @@ enum
...
+	be_t<s32> dataType;
+	be_t<s32> sampleRate;
+};
+
+struct CellVoicePortParam
+{
+	be_t<s32> portType;
+	be_t<u16> threshold;
+	be_t<u16> bMute;
+	be_t<float> volume;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.h]: f32 for consistency.

**Comment by leskal** [comment]
https://stackoverflow.com/questions/25542390/anonymous-union-and-struct/25543147#25543147

**Comment by Kravickas** [comment]
it just segfault Test Driver Unlimited earlier than master 😆

**Comment by isJuhn** [comment]
TDU2 segfault fixed

**Comment by isJuhn** [comment]
Updated with more stuff, big thanks to @Megamouse for providing me his WIP work on cellVoice, it really fleshed out the stubs. I think I've done the right thing with event queues now as well.

**Comment by ScorchEmber256** [comment]
Regressions in Call of Duty Black Ops 2 and Little Big Planet 2. Different configurations yield the same results. 

Little Big Planet 2:
E {PPU[0x1000000] Thread (main_thread) [0x0060d304]} cellSysutil: Filesystem mounting failed.
(8001002b)
E {dialog input thread} RSX: Pad handler expected but none initialized!
E {dialog input thread} RSX: Dialog input loop exited with error code=-3
capture
log: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2497812/RPCS3.log.gz)

Black Ops 2:
F {PPU[0x1000003] Thread (Block1) [0x002fe1cc]} MEM: Access violation writing location 0xd0029d90
F {PPU[0x100000a] Thread (hang detect) [0x002fe1f4]} MEM: Access violation writing location 0xd0071d80
log: [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2497817/RPCS3.zip)

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 5]
```diff
@@ -1,33 +1,143 @@
...
 
-s32 cellVoiceCreatePort()
+error_code cellVoiceCreatePort(vm::ptr<u32> portId, vm::cptr<CellVoicePortParam> pArg)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceCreatePort(portId=*0x%x, pArg=*0x%x)", portId, pArg);
+
+	if (!pArg)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: The first check in this function should be for LIBVOICE_NOT_INIT, this is usually the same for all of the libraries.

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 6]
```diff
@@ -37,177 +147,616 @@ s32 cellVoiceDisconnectIPortFromOPort()
...
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
+	if (!bitrate)
+	{
+		return CELL_EFAULT;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: There is no check for this, nullptr simply access violates.

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 7]
```diff
@@ -37,177 +147,616 @@ s32 cellVoiceDisconnectIPortFromOPort()
...
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
+	if (!bMuted)
+	{
+		return CELL_EFAULT;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: There is no check for this, nullptr simply access violates.

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 8]
```diff
@@ -37,177 +147,616 @@ s32 cellVoiceDisconnectIPortFromOPort()
...
 
-s32 cellVoiceGetPortAttr()
+s32 cellVoiceGetPortAttr(u32 portId, s32 attr, vm::ptr<void> attrValue)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceGetPortAttr(portId=%d, attr=*0x%x, attrValue=*0x%x)", portId, attr, attrValue);
+
+	if (!attrValue)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: check for LIBVOICE_NOT_INIT first.

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 9]
```diff
@@ -37,177 +147,616 @@ s32 cellVoiceDisconnectIPortFromOPort()
...
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
+	if (true)
+	{
+		return CELL_VOICE_ERROR_DEVICE_NOT_PRESENT;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: needs proper description, the intention is unclear at first sight.

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 10]
```diff
@@ -37,177 +147,616 @@ s32 cellVoiceDisconnectIPortFromOPort()
...
 
-s32 cellVoiceGetSignalState()
+s32 cellVoiceGetSignalState(u32 portId, s32 attr, vm::ptr<void> attrValue)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceGetSignalState(portId=%d, attr=%d, attrValue=*0x%x)", portId, attr, attrValue);
+
+	if (!attrValue)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: checked IDA, this check doesnt exist for this function.

File: rpcs3/Emu/Cell/Modules/cellVoice.cpp [hunk 11]
```diff
@@ -37,177 +147,616 @@ s32 cellVoiceDisconnectIPortFromOPort()
...
 
-s32 cellVoiceInit()
+error_code cellVoiceInit(vm::ptr<CellVoiceInitParam> pArg)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceInit(pArg=*0x%x)", pArg);
+
+	if (!pArg)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellVoice.cpp]: wrong order as mentioned in other comments, LIBVOICE_INITIALIZED should be the first.

**Comment by DumF0rGaming** [comment]
@isJuhn was it possible to see the text and menus back in early days of this PR development? I have compiled 0.0.5-7493-7bf267c9 Alpha with it and I have all fonts broken or missing, menu backgrounds are shown partially only.

**Comment by uaqlover** [comment]
any updates ?

**Pull request closed without merging**